src/libDirectoryService/FinalBlockPostProcessing.cpp

```diff
@@ -184,8 +184,30 @@ void DirectoryService::SendFinalBlockToShardNodes(
                 + 1
             << "] FBBLKGEN");
 
-        P2PComm::GetInstance().SendBroadcastMessage(shard_peers,
-                                                    finalblock_message);
+        if (BROADCAST_GOSSIP_MODE)
```

Comment:
Hmm if we have gossip enabled, then the code block at line 143 is partially useless and the log messages in it misleading.  I think we can optimize this by avoiding going through a temp shard_peers if gossip is enabled. Also, we only need to log the actual peers stored inside the shardFinalBlockReceivers for the gossip case.